### ISSUE TITLE
How to plot confusion matrix in yolov5-cls

### ISSUE THREAD

**Issue by vlad1mirJ** [ultralytics/yolov5#12826]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello! I am trying to find out how to plot confusion matrix for yolov5 classification task. I want to compare confusion matrix together with other metrics of trained models on custom dataset of both yolov5 and yolov8. Any ideas , suggestions or help would be appreciated!

**Comment by glenn-jocher** [comment]
@vlad1mirJ hello! Thanks for reaching out and great question. 😄

To plot a confusion matrix in your YOLOv5 classification task, after training your model, you can use the `--conf-mat` flag in the `val.py` script. It will generate a confusion matrix for your validation dataset. Here's a simple example on how to do it:

```bash
python val.py --weights your_trained_model.pt --data your_data.yaml --task val --conf-mat
```

This will produce a confusion matrix plot among other metrics in your run's results folder. Remember, visualizing performance metrics like a confusion matrix is a great way to understand your model's behavior across different classes.

Don't hesitate to dive into the [documentation](https://docs.ultralytics.com/yolov5/) for more details on custom training and evaluation.

Wishing you success with your YOLOv5 and YOLOv8 models! If you have further questions or need more assistance, we're here to help. 🚀

**Comment by vlad1mirJ** [comment]
@glenn-jocher Thanks for your fast response! It looks to me that you are describing it for yolov5 detection model rather than classification. For classification there is no need in `my_data.yaml`. Also, there is a separate script for classification validation, but it does not produce a confusion matrix. Please correct me if I am wrong 😄

**Comment by glenn-jocher** [comment]
@vlad1mirJ My apologies for the confusion earlier! 😅 You're absolutely right, for classification tasks, the approach differs slightly from detection tasks.

For YOLOv5 classification, we currently don't have a built-in option to directly generate a confusion matrix through the validation script. However, you can obtain the predictions and then use a python library like `sklearn` to compute and plot the confusion matrix. Here's a quick example assuming you have your predictions and true labels:

```python
from sklearn.metrics import confusion_matrix
import seaborn as sns
import matplotlib.pyplot as plt

# true_labels = [...] # Your actual labels
# preds = [...] # Your model predictions

conf_mat = confusion_matrix(true_labels, preds)
sns.heatmap(conf_mat, annot=True, fmt='d')
plt.xlabel('Predicted labels')
plt.ylabel('True labels')
plt.show()
```

This method requires you to manually collect your true labels and predictions after running the model. I hope this helps, and feel free to reach out if you have more questions or need further clarification. Keep up the great work! 🚀

**Comment by vlad1mirJ** [comment]
@glenn-jocher Thank you for your support! It's closer to what I have been looking. I guess the only question left is how do I obtain `true_labels` and `preds` in validation script. I had a look through the script and it seems that these values is what I am looking for. https://github.com/ultralytics/yolov5/blob/db125a20175384d75560dc9af7fb1100d67213fe/classify/val.py#L105  Could you confirm it ? If thats not the case, could you please suggest on a way to get true_labels and predictions in validation script.

**Comment by glenn-jocher** [comment]
@vlad1mirJ You've got it! That part of the code indeed helps you gather the necessary data for `true_labels` and `preds`. After your model has made predictions during validation, you can extract the true labels and predictions from the output.

Here's a succinct example to illustrate extracting these variables right after your model makes predictions:

```python
# assuming 'targets' and 'preds' are your true labels and predictions respectively
true_labels = targets.numpy()  # Converting tensors to numpy arrays if they aren't already
predictions = preds.numpy()

# Now, you can use 'true_labels' and 'predictions' for further analysis or plotting the confusion matrix as shown previously.
```

Just make sure to adjust variable names as per your actual code context. This snippet is under the assumption that you have access to both `targets` and `preds` within the script where you intend to analyze or visualize your results.

Feel free to get back if there's more to explore or clarify. Happy coding! 🚀
